### PR TITLE
ARROW-8430: [CI] Configure self-hosted runners for Github Actions

### PR DESCRIPTION
**Pull request by kszucs**
Findings:
- we require recent git version >= 2.18
- no specific way to detect whether the job runs in a self hosted runner
- to use runner specific environment variable we can setup .bashrc but that requires an interactive bash session
- the setup-python action downloads prebuilt python distributions and only the official gha virtual environments are going to be supported
- we can use the preinstalled python on the runner but to setup archery we need to add ~/.local/bin to the path using the bashrc workaround described above
- in order to have the github actions cache working we cannot use named docker volumes but once we switch to ordinary docker volumes we have the permission issue where the mounted directories are written by the root user (on linux at least) ~, so I had to pass the docker user to match the host user and group, but then we don't have enough permission to create the /build directory within the container so I had to change the widely used build directory to /tmp~
- ~need to initialize `.docker` directory for volume caches before docker does with root permissions~

### CODE REVIEW

**Comment by pitrou** [comment]
For the record, rather than adding more jobs, I think it may be more worthwhile to move some existing jobs to self-hosted.
(we're going to want to move a couple jobs from AppVeyor to GHA as well, but we don't want GHA to become capacity-limited too much)

**Comment by kszucs** [comment]
> For the record, rather than adding more jobs, I think it may be more worthwhile to move some existing jobs to self-hosted.

This PR is mostly about making the builds configurations compatible with the self-hosted runners, we can easily copy around these configs later.

> (we're going to want to move a couple jobs from AppVeyor to GHA as well, but we don't want GHA to become capacity-limited too much)

We should not worry until our feedback time get worse, then we can set low priority builds as cron jobs or do some action then.

**Comment by kszucs** [comment]
Finally it seems to work now with caching enabled (we probably need to tune the images though). 

@kou it might be interesting for you.

**Comment by kou** [comment]
Thanks for notifying me!
I'll take a look this in a few days.

**Comment by kszucs** [comment]
## Docker permission issue on linux

By default root is used within the container, thus the volumes get written as root.

The only way to create docker volumes with the host user's permissions is to use the same user (uid:gid) within the container. This is possible with `docker run --user==uid:gid` but that has two issues:

1. the user doesn't exists in the container
    SOLUTION:
        create the user and group during builds time and grant right permissions to write system resources (like /usr or /opt)
    TRADEOFF:
        it would render the cached docker layers unportable, meaning that each developer would need to rebuild the image locally to have the same uid gid pairs as on her/his host

2. the user (even if it would exist) doesn't have the right permissions
    SOLUTION:
        Create a directory writable by anyone and use that for ccache, building and installing.
    TRADEOFF:
        We cannot test the system installation, we would need to update almost all of the build scripts (and there could be unforseen issues because of the unnamed/unexistent user and group)

The easiest way to use docker user namespaces (there can be a single one set up) configured on the docker daemon which has its own limitations, but basically maps a host user:group to another user:group within the container. As an example the folders written by the root user from within the container looks like they were written by another user on the host.

Note that it only affects docker on linux because on mac and windows it works as expected (non-native containerization).

File: .github/workflows/cpp_cron.yml [hunk 1]
```diff
@@ -95,12 +95,13 @@ jobs:
       - name: Setup Python
         uses: actions/setup-python@v1
         with:
-          python-version: 3.8
+          python-version: 3.6
```
[the remark this answers is not in the document]
Comment (kszucs) [reply .github/workflows/cpp_cron.yml:98]: Yes.

**Comment by kszucs** [comment]
@kou I'm afraid that CMAKE_BUILD_PARALLEL_LEVEL doesn't propagate well with ninja, so it causes OOM during compilation in the arm64v8 build.

**Comment by kou** [comment]
Ah, `CMAKE_BUILD_PARALLEL_LEVEL` is supported since CMake 3.12 but Ubuntu 18.04 ships CMake 3.10.
Can we use Ubuntu 20.04 instead of Ubuntu 18.04?

**Comment by kszucs** [comment]
Sure, updating to use never ubuntu versions.

File: .github/workflows/cpp.yml [hunk 2]
```diff
@@ -92,6 +92,67 @@ jobs:
...
+    #         actions team, so python>3.6 must be preinstalled on the self
+    #         hosted machines
+    name: ${{ matrix.title }}
+    runs-on: ${{ matrix.runner }}
+    # if: ${{ !contains(github.event.pull_request.title, 'WIP') && github.repository == 'apache/arrow' }}
+    defaults:
+      # to use certain environment variables are set by .bashrc an interactive
+      # bash shell must be used
+      run:
+        shell: bash -i {0}
```
[the remark this answers is not in the document]
Comment (kszucs) [reply .github/workflows/cpp.yml:111]: I'll change it in a follow-up.

**Comment by kszucs** [comment]
Once approved I'll change the arm jobs to be triggered on push instead on pull request.

File: ci/docker/ubuntu-14.04-cpp.dockerfile [hunk 3]
```diff
@@ -68,7 +68,7 @@ ENV ARROW_BUILD_TESTS=ON \
     ARROW_WITH_BROTLI=ON \
     ARROW_WITH_BZ2=ON \
     ARROW_WITH_LZ4=ON \
-    ARROW_WITH_OPENSSL=OFF \
+    PARQUET_REQUIRE_ENCRYPTION=OFF \
```
Comment (kou) [inline ci/docker/ubuntu-14.04-cpp.dockerfile:71]: Could you sort in alphabetical order?

File: ci/scripts/cpp_build.sh [hunk 4]
```diff
@@ -127,7 +127,11 @@ cmake -G "${CMAKE_GENERATOR:-Ninja}" \
       ${CMAKE_ARGS} \
       ${source_dir}
 
-time cmake --build . --target install
+if [ ! -z "${CPP_MAKE_PARALLELISM}" ]; then
+  time cmake --build . --target install -- -j${CPP_MAKE_PARALLELISM}
```
[the remark this answers is not in the document]
Comment (kou) [reply ci/scripts/cpp_build.sh:131]: Does `CMAKE_BUILD_PARALLEL_LEVEL` not work with Ubuntu 20.04?

**Review by kou — APPROVED** [review]

**Comment by kszucs** [comment]
Tried but it didn’t work. I can further investigate it, but passing “-j” fixed the resource issues for now.

**Comment by kszucs** [comment]
+1, merging on green

**Pull request closed without merging**